File: Data Structures & Algorithms/valid-palindrome-ii/submission-4.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1

        while left < right:
            if s[left] != s[right]:
                return self._check(s, left + 1, right) or \
                       self._check(s, left, right - 1)
            left += 1
            right -= 1

        return True

    def _check(self, s, l, r):
        while l < r:
            if s[l] != s[r]:
                return False
            l += 1
            r -= 1
        return True

        

File: Data Structures & Algorithms/minimum-recolors-to-get-k-consecutive-black-blocks/submission-4.js
class Solution {
    /**
     * @param {string} blocks
     * @param {number} k
     * @return {number}
     */
    minimumRecolors(blocks, k) {
        let currentWhites = 0
        
         // 1. Count 'W's in the very first window of size k
        for(let i=0; i<k; i++){
            if(blocks[i] === "W"){
                currentWhites++
            }
        }

        let minOps = currentWhites;
         // 2. Slide the window across the rest of the string
        for(let i=k; i<blocks.length; i++){
            // Add the new element entering the right side of the window
            if (blocks[i] === "W") currentWhites++

            // Remove the old element leaving the left side of the window
            if(blocks[i - k] === "W") currentWhites--

            minOps = Math.min(minOps, currentWhites)
        }


        return minOps;
    }
}
